chainstate/src/detail/chainstateref.rs

```diff
@@ -275,34 +274,36 @@ impl<'a, S: BlockchainStorageRead> ChainstateRef<'a, S> {
         calculate_witness_merkle_root(block.transactions()).map_or(
             Err(CheckBlockError::WitnessMerkleRootMismatch),
             |witness_merkle| {
-                if witness_merkle_root != witness_merkle {
-                    Err(CheckBlockError::WitnessMerkleRootMismatch)
-                } else {
-                    Ok(())
-                }
+                ensure!(
+                    witness_merkle_root == witness_merkle,
+                    CheckBlockError::WitnessMerkleRootMismatch
+                );
+                Ok(())
             },
         )?;
 
         match &block.prev_block_id() {
             Some(prev_block_id) => {
-                let median_time_past = self.calculate_median_time_past(prev_block_id);
-                if block.block_time() < median_time_past {
-                    // TODO: test submitting a block that fails this
-                    return Err(CheckBlockError::BlockTimeOrderInvalid);
-                }
-
-                let max_future_offset = self.chain_config.max_future_block_time_offset();
-                if i64::from(block.block_time()) > time::get() + max_future_offset.as_secs() as i64
-                {
-                    // TODO: test submitting a block that fails this
-                    return Err(CheckBlockError::BlockFromTheFuture);
-                }
+                let median_time_past = calculate_median_time_past(self, prev_block_id);
+                ensure!(
+                    block.block_time() >= median_time_past,
+                    CheckBlockError::BlockTimeOrderInvalid
+                );
```

Comment:
A very minor nitpick:

You can add a tailing comma at the end, to make it consistent with how function calls are usually formatted:
```rust
                ensure!(
                    block.block_time() >= median_time_past,
                    CheckBlockError::BlockTimeOrderInvalid,
                );
```

Sadly rustfmt cannot make any assumptions on whether the macro is able to deal with this or not so it conservatively leaves the original punctuation there.